Add GCP plugin

MikaelSmith opened the pull request:
Adds a GCP plugin with Google Compute Engine support.

Resolves #390.

MikaelSmith committed:
Add GCP plugin

Creates a new core plugin for Google Cloud Platform. Initially it only
lists available projects for the current application credentials.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
List GCE instances

For each project, lists compute instances from GCE in that project.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
This is still a work in progress. I have some basic metadata and console output working, but not really complete. And Exec - based on https://cloud.google.com/compute/docs/instances/adding-removing-ssh-keys - is going to take some time. The goal is to have it as smooth as `gcloud compute ssh`.

MikaelSmith committed:
Add gcp.projects to restrict projects to a whitelist

This can be useful when you have limited permissions on some projects
and don't want to include them.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Add support for GCP Storage

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
List all objects in a bucket using their full name

This puts all objects as a file under the bucket. Delimited prefixes -
used to declare "directories" - are included in the name and the '/'
delimiter will be replaced with '#' (as is standard in Wash).

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Switch to listing GCP Storage hierarchically

List storage prefixes and objects hierarchically using the Storage API's
Query parameters. These are explicitly designed to return results in a
directory-like fashion. This works well for buckets with large numbers
of objects, but can be slower for deeply nested buckets as we perform
two queries for every directory.

The additional query - beyond listing the directory's contents - is to
get its attributes when we name the directory. Using Query means that a
prefix - directory - only has its name returned when listing things
under that prefix. Its attributes would be returned when you queried
with that prefix, but by that point we already want to have core
attributes like creation and update time. Therefore prefixes with lots
of child prefixes may be slow.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Implement Open for GCP Storage Objects

Uses a trivial implementation for Open that doesn't perform well when
reading lots of small chunks.

Omits Stream because the Google Cloud Storage API doesn't provide an
interface that will read indefinitely. Falling back to tail's behavior
of rechecking attributes to see when size changed makes the most sense.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Implement Exec consistent with 'gcloud compute ssh'

Implements Exec with the semantics established by the `gcloud compute
ssh` command. It creates and re-uses the `google_compute_engine` key
pair and `google_compute_known_hosts` file.

Also improves configuring the SSH transport from an SSH config.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Raise activity log to info

This makes the `info` level more useful. Previously it only included
startup/shutdown messages. Debug level was too verbose, as it includes
all filesystem activity.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
I have a few outstanding items, but I think the code that's present is pretty complete
- [ ] add tests
- [ ] provide an expiration date for SSH public keys that are uploaded to a project/instance
- [ ] check that this works for instances that explicitly prohibit project-wide SSH keys

I'm also open to suggestions on things that should be configurable, with supporting use-cases.

ekinanp commented on activity/journal.go:
```diff
@@ -103,7 +103,7 @@ func (j Journal) Warnf(msg string, a ...interface{}) {
 // `wash/activity/ID.log`.
 func (j Journal) Record(msg string, a ...interface{}) {
-	log.Debugf(msg, a...)
+	log.Printf(msg, a...)
```
Won't this cause a lot of noise to the shell experience (e.g. seeing info messages when I'm doing a wash exec)?

MikaelSmith replied:
Wash the shell defaults to `warn` log level. The commit message talks about the rationale a bit more.

MikaelSmith replied:
The one area it could be messy is if we do #302, but in that case I think activity journal tracking becomes more complicated, and we need to send the logs back to the client that originated them anyway.

ekinanp replied:
Hmm, ok. When I do `./wash`, I still see INFO messages popping up when I try to do stuff (including server start up messages). This is from the latest Wash build.

MikaelSmith replied:
Do you have `loglevel: info` set in your config?

ekinanp replied:
Ah that's why. Cool, I don't have any other concerns around this.

ekinanp commented:
Could you paste the stree output here? Curious to see what this looks like.

ekinanp commented on plugin/gcp/computeDir.go:
```diff
@@ -0,0 +1,68 @@
+func (c *computeDir) List(ctx context.Context) ([]plugin.Entry, error) {
+	var entries []plugin.Entry
+	instReq := c.Instances.AggregatedList(c.projectID)
+	if err := instReq.Pages(ctx, func(instancePage *compute.InstanceAggregatedList) error {
```
Nit: Could `err :=` move out of the if statement? Bit hard to read as it is right now.

ekinanp commented on plugin/gcp/root.go:
```diff
@@ -0,0 +1,101 @@
+// Package gcp presents a filesystem hierarchy for Google Cloud Platform resources.
+//
+// It follows https://cloud.google.com/docs/authentication/production to find your credentials:
```
Isn't this already documented? Why state it more than once?

MikaelSmith replied:
Different audiences. But you're probably right that it doesn't need to be in the godocs.

ekinanp replied:
Yeah, that's one less piece of documentation to maintain.

ekinanp commented on plugin/gcp/storageBucket.go:
```diff
@@ -0,0 +1,70 @@
+		if err != nil {
+			return nil, err
+		}
+		if objAttrs.Prefix != "" {
```
A high-level comment describing these checks (and how the storage API works) would be useful, because this isn't obvious to me w/o having to dig into the package myself.

ekinanp commented on plugin/gcp/computeDir.go:
```diff
@@ -0,0 +1,68 @@
+
+type computeDir struct {
+	plugin.EntryBase
+	computeProjectService
```
Nit: Think it's better if embedding's only restricted to `EntryBase`. That way, it's easier to see that this is strictly an Entry, and that we're just passing the project service object around. The pattern's also consistent with the other core-plugins.

ekinanp commented on plugin/gcp/storageObjectPrefix.go:
```diff
@@ -0,0 +1,42 @@
+
+// Takes the name of the directory, as well as the full prefix path.
+// Attrs may be nil if they could not be retrieved. Some prefixes don't appear to have attributes.
+func newStorageObjectPrefix(bucket *storage.BucketHandle,
```
Strange that prefixes here have attributes but S3 prefixes don't.

MikaelSmith replied:
Yeah. Surprised me.

ekinanp commented on plugin/gcp/storageObjectPrefix.go:
```diff
@@ -0,0 +1,42 @@
+		prefix:    prefix,
+	}
+	if attrs != nil {
+		pre.Attributes().SetMtime(attrs.Updated).SetSize(uint64(attrs.Size)).SetMeta(attrs)
```
What does Size mean here? Kind of weird seeing it set w/o implementing Readable.

MikaelSmith replied:
Usually 0. Directories do not universally have size 0 however. https://unix.stackexchange.com/a/77. I would interpret it as the size of the readdir response.

ekinanp reviewed (commented):
Left some initial comments. I'll review the exec stuff tomorrow morning.

ekinanp commented on transport/ssh.go:
```diff
@@ -39,51 +40,139 @@ func newAgent() (ssh.AuthMethod, error) {
+		}
+
+		// Retry until period expires or connection is established.
+		if retries > 0 {
```
Nit: You could also do `if retries <= 0 { retries = 1 }` to centralize this code.

MikaelSmith replied:
I don't think that would have the same meaning.

MikaelSmith commented:
stree output (after I fixed it)
```
gcp
└── [project]
    ├── compute
    │   └── [instance]
    │       └── console.out
    └── compute
        └── [bucket]
            ├── [prefix]
            │   ├── [prefix]
            │   └── [object]
            └── [object]
```

ekinanp reviewed (commented):
Overall Exec looks good. Just some comments.

MikaelSmith commented on plugin/gcp/computeInst.go:
```diff
@@ -36,3 +49,244 @@ func (c *computeInstance) ChildSchemas() []*plugin.EntrySchema {
+var errKeyAdded = fmt.Errorf("Key added to metadata")
+
+// SSH public keys are added to Common Instance Metadata as described in
+// https://cloud.google.com/compute/docs/instances/adding-removing-ssh-keys
```
> Adding the key to metadata in the prescribed format (mapping user to public key content) causes it to be added to `~/.ssh/authorized_keys` for the supplied user on associated instances.

I'll add it as a comment.

MikaelSmith commented on plugin/gcp/computeInst.go:
```diff
@@ -36,3 +49,244 @@ func (c *computeInstance) ChildSchemas() []*plugin.EntrySchema {
+}
+
+func generateKeys(publicKeyPath, privateKeyPath string) error {
+	privateKey, err := rsa.GenerateKey(rand.Reader, 4096)
```
I'm not sure what you mean. It's pretty much the only clear way to do it: generate an RSA key, serialize the public and private portions, write to file. There's no higher level abstraction for it.

ekinanp replied:
Cool, then the comment's unnecessary. The reason I asked the Q was b/c there's a bunch of crypto stuff out there (most of which I don't know) that I was thinking maybe there was another way to generate these keys.

MikaelSmith commented:
Updated, but I still have some TODOs to get to.

MikaelSmith commented:
Pushed an update to revamp how we check and update SSH keys so it works with more configurations.

MikaelSmith commented:
The key expiration doesn't do what I hoped it would (remove the key from the VM when it expires), so I left it out.

MikaelSmith committed:
Add volume.FS to GCE instances

Enables viewing their filesystem.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Create MetadataJSONFile helper for exposing metadata as a JSON file

This is a common pattern, so provide a helper in the `plugin` module
and use it for AWS, Docker, and GCP.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

ekinanp reviewed (approved):
Comment's not a PR blocker.

MikaelSmith commented on plugin/metadataJSON.go:
```diff
@@ -0,0 +1,54 @@
+		other:     other,
+	}
+
+	if other.getTTLOf(MetadataOp) == -1 {
```
It's not possible to set negative TTL via CachedOp. So it'd only be 0, which means "use the cache's default". DisableCachingFor sets to a negative (-1). So I'll change it to `< 0`, and add a comment about what 0 means for CachedOp.
